scertecd/scertecd.go

```diff
@@ -0,0 +1,838 @@
+//   - {prod,dev}/certd/https-cert/{domain-name}/ecdsa: PEM of private key, domain cert, LetsEncrypt cert
+package scertecd
+
+import (
+	"bytes"
+	"context"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"html"
+	"io"
+	"log"
+	"net/http"
+	"os"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/route53"
+	"github.com/tailscale/setec/client/setec"
+	"github.com/tailscale/setec/types/api"
+	"golang.org/x/crypto/acme"
+)
+
+// Server is the scertec updater server.
+//
+// Despite the name "server", it can also be used in a single-shot
+// foreground mode via its UpdateAll method.
+//
+// All exported fields must be initialzed before calling an exported
+// method on
+type Server struct {
+	SetecClient setec.Client     // required client for setec
+	Domains     []string         // domains to maintain certs for
+	Now         func() time.Time // if nil, time.Now is used
+	ACMEContact string           // optional email address for ACME registration
+	DevMode     bool             // if true, use "dev/" prefix for setec secrets instead of "prod/"
+
+	dtsOnce sync.Once
+	dts     []domainAndType
+
+	mu       sync.Mutex
+	acLazy   *acme.Client // nil until needed via getACMEClient
+	last     map[domainAndType]*certUpdateCheck
+	secCache map[string]*api.SecretValue
+}
+
+func (s *Server) initDomainAndTypes() {
+	s.dtsOnce.Do(func() {
+		for _, d := range s.Domains {
+			for _, typ := range []CertType{RSACert, ECDSACert} {
+				s.dts = append(s.dts, domainAndType{d, typ, s.DevMode})
+			}
+		}
+	})
+}
+
+// UpdateAll checks or updates all certs once and returns.
+//
+// If all certs are either fine or successfully updated, it returns nil.
+func (s *Server) UpdateAll() error {
+	s.initDomainAndTypes()
+	for _, dt := range s.dts {
+		cu := s.newCertUpdateCheck(dt)
+		st, err := cu.updateIfNeeded(context.Background())
+		if err != nil {
+			cu.lg.Printf("updateIfNeeded error: %v", err)
+			return err
+		}
+		cu.lg.Printf("success; %+v", st)
+	}
+	return nil
+}
+
+// Start starts a background renewal goroutine for each cert domain and
+// algorithm type. The context is used only for the initial ACME registration
+// check and not used thereafter.
+func (s *Server) Start(ctx context.Context) error {
+	s.initDomainAndTypes()
+	if _, err := s.getACMEClient(ctx); err != nil {
+		return err
+	}
+	for _, dt := range s.dts {
+		go s.renewCertLoop(dt)
+	}
+	return nil
+}
+
+func (s *Server) getACMEClient(ctx context.Context) (*acme.Client, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if s.acLazy != nil {
+		return s.acLazy, nil
+	}
+
+	ac, err := s.getOrMakeACMEClient(ctx, log.Printf)
+	if err != nil {
+		return nil, fmt.Errorf("getOrMakeACMEClient: %w", err)
+	}
+	if err := s.initACMEReg(ctx, log.Printf, ac); err != nil {
```

Comment:
Done, but the current was on purpose to make it impossible to log to the wrong spot. Inside the certUpdater, I didn't want it to be possible to accidentally log to the cu.s.logf instead of the cu.logf that also tees into the per-certUpdate-run log buffer as visible in the web UI.

But a little discipline is fine. It cuts down on parameters. And worst case some logs don't show up in the web UI.